packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -308,8 +308,13 @@ export default util.createRule<Options, MessageId>({
         return;
       }
 
+      // Known edge case: truthy primitives and nullish values are always valid boolean expressions
+      if (is('nullish', 'truthy string') || is('nullish', 'truthy number')) {
```

Comment:
I think that these two checks should probably be guarded using the `options.allowString` and `options.allowNumber` checks.

Otherwise someone could set `allowNumber: false` and find that:
```
declare const x: 1 | null;
declare const y: 1;
if (x) {} // passes linting
if (y) {} // fails linting
```
which is a weird inconsistency